src/stubadub/core.clj

```diff
@@ -3,15 +3,20 @@
 (def ^:dynamic *calls*)
 
 (defmacro with-stub [func & opts-and-body]
-  (let [has-return-value? (= :returns (first opts-and-body))
-        body (if has-return-value? (nnext opts-and-body) opts-and-body)
-        return-value (when has-return-value? (second opts-and-body))]
-    `(binding [*calls* (if (bound? #'*calls*)
-                         *calls* (atom []))]
-       (with-redefs [~func (fn [& args#]
+  (let [has-stub? (some #{(first opts-and-body)} [:returns :calls])
+        body (if has-stub? (nnext opts-and-body) opts-and-body)
+        stub-value (when has-stub? (second opts-and-body))]
+    `(let [stub-func# (if ~has-stub?
+                        (if (fn? ~stub-value)
+                          ~stub-value
+                          (constantly ~stub-value))
+                        (constantly nil))]
```

Comment:
I think I'd like this code better if it had used `has-return-value?` like before, introducing a separate `has-return-fn?` or some such. Two reasons for that:

- `with-stub` then `has-stub?` overloads the term `stub`
- the `:returns` and `:calls` are quickly discarded. A `:returns` that returns a function would fail.